data/man/luma.1
.\" Man page generated from reStructeredText.
.
.TH LUMA 1 "April 26, 2011" "" "LDAP administration utility"
.SH NAME
luma \- LDAP browser and administration utility
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.sp
luma [OPTIONS]
.SH DESCRIPTION
.sp
Luma is a cross\-platform, open source LDAP browser and administration utility,
capable of managing data stored on LDAP enabled server. It is written in Python,
with PyQt4 (\fIGUI\fP) and python\-ldap (\fILDAP connections\fP). Luma supports plugins
through its own plugin system. A selection of plugins, providing useful LDAP
functionality, is includeed with the base application:
.INDENT 0.0
.IP \(bu 2
.
\fIBrowser\fP:
Provides a familiar interface to browse the entry tree on connected LDAP
enabled servers. Entry attributes can be edited and deleted. Supports adding
filters to limit expansions on large subtrees.
.IP \(bu 2
.
\fITemplate\fP:
.IP \(bu 2
.
\fISearch\fP:
A specialized search plugin to do advanced and complex search on a LDAP
server. Supports filter creation that can be applied in the Browser plugin.
.UNINDENT
.sp
Plugins included in \fILuma 2.4\fP, but not ported from PyQt3 yet:
.INDENT 0.0
.IP \(bu 2
.
\fISchema browser\fP:
.IP \(bu 2
.
\fIAddressbook\fP: Supports building addressbooks from LDAP entries on different
servers.
.IP \(bu 2
.
\fIMassive user creation\fP:
.IP \(bu 2
.
\fIAdmin utilities\fP:
.UNINDENT
.SH OPTIONS
.sp
A summary of the options supported by \fIluma\fP is included below.
.INDENT 0.0
.TP
.B \-h,  \-\-help
.
Display a help message and exit.
.TP
.B \-v,  \-\-verbose
.
print more error, debug and info messages to the console.
.TP
.B \-\-clear\-config
.
clear the config file before launching Luma.
.TP
.B \-\-clear\-serverlist
.
clear the serverlist before launching Luma.
.TP
.B \-\-clear\-templates
.
clear the templates file before launching Luma.
.TP
.B \-\-clear\-filters
.
clear the filters file before launching Luma.
.TP
.B \-\-clear\-all
.
clear everything before launching Luma. This will result in the same as
providing all the before mentioned clear options.
.UNINDENT
.\" --config-dir=DIR
.
.\" run Luma with another configuration directory.
.
.\" --plugin-dir=DIR
.
.\" define another directory to look for plugins. DIR will be appended to the
.
.\" list of default plugin directories.
.
.SH FILES
.INDENT 0.0
.TP
.B \fB~/.config/luma\fP
.sp
Default directory for configuration data.
.TP
.B \fB~/.config/luma/luma.conf\fP
.sp
Application settings.
.TP
.B \fB~/.config/luma/serverlist.xml\fP
.sp
The serverlist.
.TP
.B \fB~/.config/luma/templatelist.xml\fP
.sp
The templatelist created by the Template plugin.
.TP
.B \fB\-/.config/luma/filters\fP
.sp
The saved LDAP filters created by the Search plugin.
.UNINDENT
.SH BUGS
.sp
Bug tracker: \fI\%http://sourceforge.net/tracker/?group_id=89105\fP
.SH RESOURCES
.sp
Website: \fI\%http://luma.sf.net\fP
.SH AUTHORS
.sp
Originally written and developed by Wido Depping \fI\%wido@users.sourceforge.net\fP
from 2003, with help from Bjørn Over Grøtan \fI\%bgrotan@grotan.com\fP and Vegar
Westerlund \fI\%vegarwe@users.sourceforge.net\fP.
.sp
Rewritten for PyQt4 and ported to multiple platforms in 2011 by Christian
Forfang \fI\%cforfang@gmail.com\fP, Einar Uvsløkk \fI\%einar.uvslokk@gmail.com\fP, Johannes
Harestad \fI\%johannesharestad@gmail.com\fP, Per Ove Ringstad \fI\%peroveri@stud.ntnu.no\fP,
Simen Natvig \fI\%simen.natvig@gmail.com\fP, Sondre Frisvold \fI\%sondre.frisvold@c2i.net\fP
.SH COPYRIGHT
GNU General Public License version 2 or later
.\" Generated by docutils manpage writer.
.\" 
.


ChangeLog
*********
ChangeLog
*********
.. This changelog will mainly contain changes made between different tags in
.. the master branch. Changes made in the devlopment branch is documented in
.. the git log.

3.0.7 (2011-05-26)
==================

- Fixed unconsistent license statements in the sourcecode, meaning license
  boilerplates is now added/updated (from GPL v3 to GPL v2) in all files
  added and/or modified since Luma 2.4.

- Removed the last string exception from ObjectClassAttributeInfo.py!
| http://sourceforge.net/support/tracker.php?aid=3014815
| http://bugs.debian.org/585248

- Gzipping of Nroff manpages on UNIX and UNIX-like systems is now handled during
  installation.

- Include the nthash patch, which was fixed in Luma 2.4 svn rev 841, but not 
  included in the new codebase.

| http://sourceforge.net/support/tracker.php?aid=2309281
| http://bugs.debian.org/559078
| http://luma.svn.sourceforge.net/viewvc/luma?view=rev&revision=841
| https://launchpad.net/bugs/297148
| http://launchpadlibrarian.net/19590144/luma_2.3-1.1ubuntu1.debdiff

- Added a import check for the crypt module in mkpasswd, because of ImportError
  on the Windows platform.

3.0.6-2 (2011-05-16)
====================

- Fixed exception.ErrorType bug in LumaConnectionWrapper.

- Updated the error/exception feedback in the Search plugin

- Renamed the browser_plugin folder to browser

- Moved Plugin* related code out from the util package and into base.gui, 
  base.gui.design, base.model

- Fixed export subtree in Browser plugin.

3.0.6-1 (2011-05-15)
====================

- Some small fixes to the documentation files.

3.0.6 (2011-05-15)
==================
.. This marks the end of the project. Einar Uvsløkk <einar.uvslokk@linux.com>

- New Connection API introduced with the LumaConnectionWrapper. This includes
  wrapper methods (both sync and async) for the refactored LumaConnection.
  Uses signal and slots to notify results from the various operations.
  All Qt 4 dependencies is thus removed from the backend modules.

- Refactor cleanup of the menubar in the main window.

- Search plugin now supports none-ascii characters in both server name and 
  search filter strings.

- Dynamic translation implemented in the Template plugin.

- Improved (simplified) internationalization support

3.0.5-sprint5 (2011-04-11)
==========================
.. This marks the end of sprint 5. Einar Uvsløkk <einar.uvslokk@linux.com>

- Search plugin has got a few updates and improvements. The filter builder
  has replaced the old filter wizard, some UI fixes in the search form, and 
  the result view for displaying search results is implemented. New feature
  enabling additional filtering on returned search results.

- Template plugin improved with the option to define none required attributes
  to appear as required.


3.0.4-sprint4 (2011-03-25)
==========================
.. This marks the end of sprint 4. Einar Uvsløkk <einar.uvslokk@linux.com>
	
- Main-application: Support for commandline arguments added. As of now this
  includes mostly clear flags, to wipe clean malicious config settings.

- Browser-plugin: Added multiselection item export and delete. Lots of 
  improvments in the Browsers entry view, including basic support for different
  html layouts for the view.

- Template-plugin: Gui finished, logic on its way.

- Search-plugin: Skeleton gui, and basic functionality implemented.


3.0.3-sprint3 (2011-03-14)
==========================
.. This marks the end of sprint 3. Einar Uvsløkk <einar.uvslokk@linux.com>

- Luma is now able to load plugins from the default plugin location.

- The Browser plugin now supports editing entries

- A simple installation procedure is included, making use of pythons own 
  distutils module.


3.0.2-sprint2 (2011-03-01)
==========================
.. This marks the end of sprint 2. Einar Uvsløkk <einar.uvslokk@linux.com>

- Plugin support is partly implemented. Loading plugins is working, but no 
  *real* plugins is present at runtime yet.

- The Browser plugin is partly implemented, but not integrated into the
  application yet.

- An extended QSettings class is implemented, with transparent support for 
  loading and saving application  settings. Settings is loaded and saved from 
  both the main window and the newly implemented settings dialog.

- Full translation is available for English and Norwegian, and can be changed,
  from both the settings dialog and the application menu bar, at runtime 
  without changing the application state.


3.0.1-sprint1 (2011-02-28)
==========================
.. This marks the end of sprint 1. *Einar Uvsløkk* <einar.uvslokk@linux.com>

.. note::
   This relase branch was tagged and created at the end of sprint 2, but still
   reflects the work done up until the end of sprint 1. It was created to 
   document the development progression.

- The Menu bar is restructured compared to Luma 2.4, various platforms *Human
  Interface Guidelines*.

- The Plugin toolbar is implemented, but with no real functionality yet.

- The Logger Window is implemented and displayes various logging levels, using
  the builtin python logging module.

- About dialog is implemented, and completely rewritten compared to Luma 2.4,
  inspired by gtk and gnome.

- The Server dialog GUI is implemented. Support for reading and saving the 
  server list is also implemented.

- Translation files is loaded dynamically. Uses the old translation files, 
  which still works on parts of the application. The languages can be selected
  from the menu bar, which is populated dynamically on startup.



luma/luma.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# luma
#
# Copyright (c) 2011
#     Einar Uvsløkk, <einar.uvslokk@linux.com>
#     Christian Forfang, <cforfang@gmail.com>
#
# Copyright (c) 2003, 2004, 2005
#     Wido Depping, <widod@users.sourceforge.net>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see http://www.gnu.org/licenses/

import logging
import optparse
import os
import StringIO
import sys
import traceback

failed = 0

if sys.version_info < (2,6):
    failed = 1
    sys.stderr.write("""
###########################################################
## Luma needs Python version 2.6 or higher.              ##
##                                                       ##
## Please consider an upgrade.                           ##
###########################################################
""")

if sys.version_info >= (3,):
    failed = 1
    sys.stderr.write("""
###########################################################
## Luma does not yet run on Python 3.                    ##
##                                                       ##
## Please consider a downgrade.                          ##
###########################################################
""")
try:
    from PyQt4.QtCore import (QEvent, Qt)
    from PyQt4.QtGui import (QApplication, QIcon)
except ImportError, e:
    print e
    failed = 1
    sys.stderr.write("""
###########################################################
## ImportError: Unable to import module: PyQt4           ##
##                                                       ##
## PyQt4 is needed for the Graphical User Interface, and ##
## must be installed in order to successfully run Luma.  ##
## PyQt4 can be obtained from:                           ##
##                                                       ##
## http://www.riverbankcomputing.com/software/pyqt/intro ##
###########################################################
""")

try:
    import ldap
except ImportError:
    failed = 1
    sys.stderr.write("""
###########################################################
## ImportError: Unable to import module: ldap            ##
##                                                       ##
## python-ldap is needed to successfully run Luma.       ##
## python-ldap can be obtained from:                     ##
##                                                       ##
## http://python-ldap.org/                               ##
###########################################################
""")

if failed:
    print "Exiting ..."
    sys.exit(1)

del failed
import __init__ as appinfo
from base.backend.Log import LumaLogHandler
from base.gui.SplashScreen import SplashScreen
from base.gui.Settings import Settings
from base.util.Paths import getConfigPrefix

# This import ensures that all compiled resources are available for the
# running application. It contains all icons and translation files
import resources


class Luma(QApplication):
    """Possibly to be used later.
    """

    def __init__(self, argv):
        QApplication.__init__(self, argv)
        self.progressBar = None

    def setProgressBar(self, bar):
        self.progressBar = bar

    def event(self, event):
        if event.type() == QEvent.User:
            event.accept()
            self.setBusy(True)
            return True
        else:
            return QApplication.event(self, event)

    def setBusy(self, status):
        print "setBusy()"
        if status:
            self.setOverrideCursor(Qt.WaitCursor)
            if self.progressBar != None:
                self.progressBar.setRange(0, 0)
        else:
            self.restoreOverrideCursor()
            if self.progressBar != None:
                self.progressBar.setRange(0, 100)


class TempLogHandler(logging.Handler):
    def __init__(self):
        logging.Handler.__init__(self)
        self.logList = []

    def emit(self, record):
        self.logList.append(record)


def startApplication(argv, verbose=False, clear=[], dirs={}):
    """Preparing Luma for take-off

    Parameters:

    - `verbose`: boolean value indicating whether or not to print more
      than error messages to console.
    - `clear`: a list containing what should be cleared before start.
    - `dirs`: a dictionary containing containing possible dirs to
      consider on start-up.
    """
    app = Luma(argv)

    """ Fixed but not removed in case we change our minds """
    #import platform
    #if sys.platform.lower().startswith('win'):
        # Avoids ugly white background
        #from PyQt4.QtGui import QStyleFactory
        #QApplication.setStyle(QStyleFactory.create("plastique"))
        #QApplication.setPalette(QApplication.style().standardPalette())

    app.setOrganizationName(appinfo.ORGNAME)
    app.setApplicationName(appinfo.APPNAME)
    app.setApplicationVersion(appinfo.VERSION)
    app.setWindowIcon(QIcon(':/icons/128/luma'))

    # Setup the logging mechanism
    l = logging.getLogger()
    l.setLevel(logging.DEBUG)
    formatter = logging.Formatter(
        "[%(threadName)s] - %(name)s - %(levelname)s - %(message)s"
    )

    # Keep all logs from now until the GUI-LoggerWidget is up and can
    # be populated
    tmpLH = TempLogHandler()
    l.addHandler(tmpLH)

    settings = Settings()
    # Because we use QSettings for the application settings we
    # facilitate QSettings if the user wishes to start Luma fresh
    if 'config' in clear:
        clear.remove('config')
        settings.clear()

    if settings.configPrefix == '':
        # This will be the case on the first run, or if the user
        # have startet the application with the --clear-config option
        # We therefore need to retrive the config prefix in a best
        # practize cross-platform way
        #
        # NOTE: The first return value is a bool that indicates wheter
        #       the configprefix exists or not. As of now the config
        #       prefix is return regardless if it is writable or not.
        (_, settings.configPrefix) = getConfigPrefix()

    (_, configPrefix) = getConfigPrefix()

    if verbose:
        """If verbose mode is enabled we start logging to the console
        TODO: Add support for adjusting the level of verbosity ?
        """
        consoleHandler = logging.StreamHandler()
        consoleHandler.setFormatter(formatter)
        l.addHandler(consoleHandler)

    __handleClearOptions(configPrefix, clear)

    # Initialize the splash screen
    splash = SplashScreen()
    splash.show()

    # Initialize the main window

    from base.gui.MainWindow import MainWindow
    mainwin = MainWindow()

    # Set up logging to the loggerwidget
    llh = LumaLogHandler(mainwin.loggerWidget)
    l.removeHandler(tmpLH)  # Stop temp-logging
    l.addHandler(llh)  # Start proper logging

    # Populate the loggerWidget with the saved entries
    for x in tmpLH.logList:
        llh.emit(x)

    app.lastWindowClosed.connect(mainwin.close)

    mainwin.show()
    splash.finish(mainwin)

    # Add a exception hook to handle all
    # exceptions missed in the main application
    sys.excepthook = unhandledException

    # Need to activate the mainwindow in order to have focus,
    # if the application is started in fullscreen mode
    mainwin.activateWindow()
    sys.exit(app.exec_())


def main(argv):
    """Set up and parse the command line for supported application
    options and arguments.
    """
    usage = u'%prog [options] [args]'
    p = optparse.OptionParser(usage=usage)
    p.add_option(
        '-v', '--verbose',
        dest='verbose',
        action='store_true',
        help='print more error, debug and info messages to console'
    )
    p.add_option(
        '--clear-config',
        dest='clear_config',
        action='store_true',
        help='clear the config file before launching Luma'
    )
    p.add_option(
        '--clear-serverlist',
        dest='clear_serverlist',
        action='store_true',
        help='clear the serverlist before launching Luma'
    )
    p.add_option(
        '--clear-templatefile',
        dest='clear_templatefile',
        action='store_true',
        help='clear the templates file before launching Luma'
    )
    p.add_option(
        '--clear-filters',
        dest='clear_filters',
        action='store_true',
        help='clear the filters file before launching Luma'
    )
    p.add_option(
        '--clear-all',
        dest='clear_all',
        action='store_true',
        help='clear everything before launching Luma. \n ' +
        'This will result in the same as providing all the before ' +
        'mentioned clear options'
    )
    # TODO: These last two options is put here as a reminder. Would have
    #       been a cool feature to be able to launch Luma with different
    #       configuration locations.
    #           The option to specify yet another plugin directory to look
    #       for plugins could also be cool. Especially if you are in the
    #       midle of developing a new plugin, a want to do some occasionally
    #       testing.
    #
    #p.add_option(
    #    '--config-dir',
    #    dest='config_dir',
    #    action='store',
    #    type='string',
    #    metavar='DIR',
    #    help='run Luma with another configuration directory'
    #)
    #p.add_option(
    #    '--plugin-dir',
    #    dest='plugin_dir',
    #    action='store',
    #    type='string',
    #    metavar='DIR',
    #    help='define another directory to look for plugins. DIR will be ' +
    #    'appended to the list of default plugins directories'
    #)

    (opt, _) = p.parse_args()

    dirs = {}
    clear = []

    # TODO: If we append the filenames to the clear list, it becomes
    #       trivial to clear the specified files. The way we do it here
    #       works only if we are sure to use the same filenames
    #       through out the application. Also, because we use QSettings
    #       for the config file we need to handle this case differently.
    #if opt.config_dir:
    #    dirs['config'] = opt.config_dir
    #if opt.plugin_dir:
    #    dirs['plugins'] = opt.plugin_dir
    if opt.clear_all:
        clear.append('config')
        clear.append('serverlist.xml')
        clear.append('templates')
    else:
        if opt.clear_config:
            clear.append('config')
        if opt.clear_serverlist:
            clear.append('serverlist.xml')
        if opt.clear_templatefile:
            clear.append('templates.xml')
        if opt.clear_filters:
            clear.append('filters')

    startApplication(argv, opt.verbose, clear, dirs)


def __handleClearOptions(configPrefix, clear=[]):
    """If the application has been started with clear options, we
    handle these before setting up the application main window
    """
    for file in clear:
        path = os.path.join(configPrefix, file)
        if os.path.isfile(path):
            f = open(path, 'w')
            f.close()
        else:
            l = logging.getLogger('luma')
            msg = '{0} couldn\'t be located in {1}'
            l.info(msg.format(file, configPrefix))


def unhandledException(etype, evalue, etraceback):
    """UnhandledException handler
    """
    tmp = StringIO.StringIO()
    traceback.print_tb(etraceback, None, tmp)
    e = """[Unhandled (handled) Exception]
This is most likely a bug. In order to fix this, please send an email to
    <luma-users@lists.sourceforge.net>
with the following text and a short description of what you were doing:
>>>\n[{0}] Reason:\n{1}\n{2}\n<<<"""
    logger = logging.getLogger('luma')
    logger.error(e.format(tmp.getvalue(), str(etype), str(evalue)))
    # Make sure the cursor is normal
    QApplication.restoreOverrideCursor()

if __name__ == '__main__':
    main(sys.argv)


# vim: tabstop=4 expandtab shiftwidth=4 softtabstop=4
